### PR TITLE
add resampler kernel

### PR DESCRIPTION
**Pull request by autoih**
Address [issue #640](https://github.com/tensorflow/addons/issues/640), resampler kernel

### CODE REVIEW

**Comment by autoih** [comment]
Hi @WindQAQ, can you help review when you are available?

File: tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or  implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#ifndef TENSORFLOW_CONTRIB_RESAMPLER_KERNELS_RESAMPLER_OPS_H_
+#define TENSORFLOW_CONTRIB_RESAMPLER_KERNELS_RESAMPLER_OPS_H_
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops.h]: Remember to rename the header guard.

File: tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                  T* __restrict__ grad_warp, const int batch_size,
+                  const int data_height, const int data_width,
+                  const int data_channels, const int num_sampling_points);
+};
+
+}  // namespace functor
+}  // end namespace addons
+}  // namespace tensorflow
+
+#endif  // TENSORFLOW_CONTRIB_RESAMPLER_KERNELS_RESAMPLER_OPS_H_
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops.h]: ditto

File: tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+#ifndef TENSORFLOW_CONTRIB_RESAMPLER_KERNELS_RESAMPLER_OPS_H_
+#define TENSORFLOW_CONTRIB_RESAMPLER_KERNELS_RESAMPLER_OPS_H_
+
+#if PLATFORM_WINDOWS
+#define __restrict__ __restrict
+#endif
+
+namespace tensorflow {
+class OpKernelContext;
+}
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops.h]: Probably using `#include "tensorflow/xxx"` instead of doing this.

File: tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops.h [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  void operator()(::tensorflow::OpKernelContext* ctx, const Device& d,
+                  const T* __restrict__ data, const T* __restrict__ warp,
+                  const T* __restrict__ grad_output, T* __restrict__ grad_data,
+                  T* __restrict__ grad_warp, const int batch_size,
+                  const int data_height, const int data_width,
+                  const int data_channels, const int num_sampling_points);
+};
+
+}  // namespace functor
+}  // end namespace addons
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops.h]: // namespace addons

File: tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops_gpu.cu.cc [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+    ::tensorflow::GpuLaunchConfig config =
+        ::tensorflow::GetGpuLaunchConfig(output_data_size, d);
+    TF_CHECK_OK(GpuLaunchKernel(
+        Resampler2DKernel<T>, config.block_count, config.thread_per_block, 0,
+        d.stream(), data, warp, output, batch_size, data_height, data_width,
+        data_channels, num_sampling_points));
+  }
+};
+
+// TODO(fviola): gcudacc fails at compile time with Eigen::half.
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops_gpu.cu.cc]: It would be great if we could also supports `Eigen::half` though it's ok for this PR.

File: tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops_gpu.cu.cc [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+                                warp, grad_output, grad_data, grad_warp,
+                                batch_size, data_height, data_width,
+                                data_channels, num_sampling_points));
+  }
+};
+
+template struct ResamplerGrad2DFunctor<GPUDevice, float>;
+
+}  // namespace functor
+}  // end namespace addons
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops_gpu.cu.cc]: // namespace addons

File: tensorflow_addons/image/resampler_ops.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+"""Python layer for Resampler."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+from tensorflow_addons.utils.resource_loader import get_path_to_datafile
+# from tensorflow_addons.custom_ops.image.cc.ops import gen_resampler_ops
```
Comment (WindQAQ) [inline tensorflow_addons/image/resampler_ops.py]: Please remove unused line.

File: tensorflow_addons/image/resampler_ops.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+_resampler_ops = tf.load_op_library(
+    get_path_to_datafile("custom_ops/image/_resampler_ops.so"))
+
+
+@tf.function
+def resampler(data, warp, name="resampler"):
```
Comment (WindQAQ) [inline tensorflow_addons/image/resampler_ops.py]: `name=None`

File: tensorflow_addons/image/resampler_ops.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+        warp_tensor = tf.convert_to_tensor(warp, name="warp")
+        # rspr = _resampler_ops.resampler(data_tensor, warp_tensor)
+        return _resampler_ops.addons_resampler(data_tensor, warp_tensor)
+
+
+@tf.RegisterGradient("Resampler")
+def _resampler_grad(op, grad_output):
+    data, warp = op.inputs
+    grad_output_tensor = tf.convert_to_tensor(grad_output, name="grad_output")
+    # rspr_grad = _resampler_ops.resampler_grad(data, warp, grad_output_tensor)
```
Comment (WindQAQ) [inline tensorflow_addons/image/resampler_ops.py]: Please remove unused line.

File: tensorflow_addons/image/resampler_ops_test.py [hunk 10]
```diff
@@ -0,0 +1,253 @@
+# pylint: disable=bad-continuation
```
Comment (WindQAQ) [inline tensorflow_addons/image/resampler_ops_test.py]: This will disable it on the whole file. Maybe you just want to disable it for a block?
```python
# pylint: disable=bad-continuation
DO SOMETHING
# pylint: enable=bad-continuation
```

File: tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops.h (hunk 1 above)
Comment (WindQAQ) [reply tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops.h]: `TENSORFLOW_ADDONS_IMAGE_KERNELS_RESAMPLER_OPS_H_`

File: tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops.h (hunk 3 above)
Comment (autoih) [reply tensorflow_addons/custom_ops/image/cc/kernels/resampler_ops.h]: Yea, thanks @WindQAQ.

File: tensorflow_addons/image/resampler_ops_test.py (hunk 10 above)
Comment (autoih) [reply tensorflow_addons/image/resampler_ops_test.py]: Thanks for pointing this out, @WindQAQ.

File: tensorflow_addons/image/resampler_ops_test.py [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+    def test_op_backward_pass_cpu_float16(self):
+        self._test_op_backward_pass(False, tf.float16, 1e-3)
+
+    def test_op_backward_pass_cpu_float32(self):
+        self._test_op_backward_pass(False, tf.float32, 1e-4)
+
+    def test_op_backward_pass_cpu_float64(self):
+        self._test_op_backward_pass(False, tf.float64, 1e-6)
+
+    def _test_op_forward_pass(self, on_gpu, dtype, tol):
```
Comment (WindQAQ) [inline tensorflow_addons/image/resampler_ops_test.py]: I doubt there might be some mistakes for this function. The operations could only be executed when `on_gpu` is `True`. To validate if GPU is available, you might need to do something like this:

```python
use_gpu = on_gpu and tf.test.is_gpu_available()
with test_utils.device(use_gpu):
    # Do something
```

File: tensorflow_addons/image/resampler_ops_test.py [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+            reference_output = np.zeros_like(outputs)
+            for batch in range(batch_size):
+                for c in range(data_channels):
+                    reference_output[batch, :, :, c] = _bilinearly_interpolate(
+                        data[batch, :, :, c], warp[batch, :, :, 0],
+                        warp[batch, :, :, 1])
+
+            self.assertAllClose(outputs, reference_output, rtol=tol, atol=tol)
+
+    def _test_op_backward_pass(self, on_gpu, dtype, tol):
```
Comment (WindQAQ) [inline tensorflow_addons/image/resampler_ops_test.py]: ditto

File: tensorflow_addons/image/resampler_ops_test.py [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+    i_c = data[y_0, x_1]
+    i_d = data[y_1, x_1]
+
+    w_a = (x_1 - x) * (y_1 - y)
+    w_b = (x_1 - x) * (y - y_0)
+    w_c = (x - x_0) * (y_1 - y)
+    w_d = (x - x_0) * (y - y_0)
+
+    samples = (w_a * i_a + w_b * i_b + w_c * i_c + w_d * i_d)
+    samples.reshape(shape)
```
[the remark this answers is not in the document]
Comment (autoih) [reply tensorflow_addons/image/resampler_ops_test.py]: Thanks @WindQAQ. Yes, I think this is correct.

File: tensorflow_addons/image/resampler_ops_test.py [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+        data = np.zeros(data_shape)
+        warp_shape = (batch_size, warp_height, warp_width, 2)
+        warp = np.zeros(warp_shape)
+        outputs = resampler(tf.constant(data), tf.constant(warp))
+
+        # pylint: disable=bad-continuation
+        with self.assertRaisesRegexp(
+                tf.errors.UnimplementedError, "Only bilinear interpolation is "
+                "currently supported."):
+            outputs
```
[the remark this answers is not in the document]
Comment (autoih) [reply tensorflow_addons/image/resampler_ops_test.py]: Thanks @WindQAQ.

File: tensorflow_addons/image/resampler_ops_test.py (hunk 11 above)
Comment (autoih) [reply tensorflow_addons/image/resampler_ops_test.py]: Thanks for the gpu related comments, @WindQAQ. I am not currently able to test wtih gpu machine.

**Comment by autoih** [comment]
Hi @WindQAQ, I block the [part](https://github.com/tensorflow/addons/blob/fdd9a4b1004765e6cb707f3c2159672225388b12/tensorflow_addons/image/resampler_ops_test.py#L240), which cannot be tested successfully. May I have your suggestions? Thank you.

File: tensorflow_addons/image/resampler_ops_test.py [hunk 15]
```diff
@@ -0,0 +1,256 @@
...
+        warp_width = 2
+        warp_height = 6
+        batch_size = 3
+
+        warp = _make_warp(batch_size, warp_height, warp_width,
+                          dtype.as_numpy_dtype)
+        data_shape = (batch_size, data_height, data_width, data_channels)
+        data = np.random.rand(*data_shape).astype(dtype.as_numpy_dtype)
+        use_gpu = on_gpu
+        if use_gpu and tf.test.is_gpu_available():
```
Comment (WindQAQ) [inline tensorflow_addons/image/resampler_ops_test.py]: ditto

**Comment by autoih** [comment]
Thanks @WindQAQ. I've updated new patch.

File: tensorflow_addons/image/__init__.py [hunk 16]
```diff
@@ -17,16 +17,17 @@
...
 
+from tensorflow_addons.image.connected_components import connected_components
 from tensorflow_addons.image.dense_image_warp import dense_image_warp
 from tensorflow_addons.image.dense_image_warp import interpolate_bilinear
 from tensorflow_addons.image.distance_transform import euclidean_dist_transform
 from tensorflow_addons.image.distort_image_ops import adjust_hsv_in_yiq
 from tensorflow_addons.image.distort_image_ops import random_hsv_in_yiq
 from tensorflow_addons.image.filters import mean_filter2d
 from tensorflow_addons.image.filters import median_filter2d
+from tensorflow_addons.image.interpolate_spline import interpolate_spline
```
Comment (WindQAQ) [inline tensorflow_addons/image/__init__.py]: Because the import order is changed, you have to modify [this line](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/sparse_image_warp.py#L190) from
```python
interpolate_spline.interpolate_spline(
```
to
```python
interpolate_spline(
```

Comment (autoih) [reply tensorflow_addons/image/__init__.py]: Thanks a lot @WindQAQ.

File: tensorflow_addons/custom_ops/image/cc/kernels/adjust_hsv_in_yiq_op.cc [hunk 17]
```diff
@@ -28,6 +28,7 @@ limitations under the License.
 #include "tensorflow_addons/custom_ops/image/cc/kernels/adjust_hsv_in_yiq_op.h"
 
 namespace tensorflow {
+
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/adjust_hsv_in_yiq_op.cc]: The last one: Could you remove this trivial change?

File: tensorflow_addons/custom_ops/image/cc/kernels/adjust_hsv_in_yiq_op.cc [hunk 18]
```diff
@@ -168,4 +168,4 @@ REGISTER_KERNEL_BUILDER(
 #endif
 
 }  // end namespace addons
-}  // namespace tensorflow
\ No newline at end of file
+}  // namespace tensorflow
```
Comment (WindQAQ) [inline tensorflow_addons/custom_ops/image/cc/kernels/adjust_hsv_in_yiq_op.cc]: Could you revert the change of this line? I mean do any change to this file actually :-) Thanks.

**Comment by autoih** [comment]
Thanks @WindQAQ. I've tried it again. It may work now. :)
